code/espurna/curtain_kingart.ino

```diff
@@ -0,0 +1,195 @@
+
+The factory default Open and Close run time for the switch is 50 seconds, and it must be set to 
+an accurate run time for smooth working. Some motors do not have the resistance stop function,
+so when the Cover/Shutter/Blind/Curtain track open or close to the maximum length, but the motor keeps on running.
+This might cause damage on the motor and the switch, it also wastes a lot of energy. In order
+to protect the motor, this switch designed with a time setting function. After setting up the run time,
+the switch will automaticly stop when the track reaches its limits. The run time setting is also helpful
+for the accurate control when manually controlling the device via the touch panel.
+
+After installing the switch and connecting the switch for the very first time:
+- First, it will automatically close the Cover/Shutter/Blind/Curtain to the maximum.
+- Press and hold the touch interface pause button for around 4 seconds until the red background
+  led lights up and starts blinking. Then, press the open touch button so start the opening process.
+- When cover is fully open, press the Open or Close button to stop the timer and save the calculated run time.
+
+To configure the device:
+- Press up/down for 5 seconds to bring device into AP mode. After pressing up/down again, device will restart in normal mode.
+*/
+
+#if KINGART_CURTAIN_SUPPORT
+
+#include <TimeLib.h> // now() definition
+
+#ifndef KINGART_CURTAIN_PORT
+#define KINGART_CURTAIN_PORT         Serial      // Hardware serial port by default
+#endif
+
+#ifndef KINGART_CURTAIN_BUFFER_SIZE
+#define KINGART_CURTAIN_BUFFER_SIZE  100         // Local UART buffer size
+#endif
+
+#define KINGART_CURTAIN_TERMINATION  '\e'        // Termination character after each message
+#define KINGART_CURTAIN_BAUDRATE     19200       // Serial speed is fixed for the device
+  
+char _KACurtainBuffer[KINGART_CURTAIN_BUFFER_SIZE];
+bool _KACurtainNewData = false;
+  
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+void _KACurtainSend(const char * tx_buffer) {
+    KINGART_CURTAIN_PORT.print(tx_buffer);
+    KINGART_CURTAIN_PORT.print(KINGART_CURTAIN_TERMINATION);
+    KINGART_CURTAIN_PORT.flush();
+}
+  
+void _KACurtainReceiveUART() {
+    static unsigned char ndx = 0;
+    while (KINGART_CURTAIN_PORT.available() > 0 && !_KACurtainNewData) {
+        char rc = KINGART_CURTAIN_PORT.read();
+        if (rc != KINGART_CURTAIN_TERMINATION) {
+            _KACurtainBuffer[ndx] = rc;
+            if (ndx < KINGART_CURTAIN_BUFFER_SIZE - 1) ndx++;
+        } else {
+            _KACurtainBuffer[ndx] = '\0';
+            _KACurtainNewData = true;
+            ndx = 0;
+        }
+    }
+}
+  
+void _KACurtainResult() {
+    if (_KACurtainNewData) {
+
+        // Need to send confiramtion to the N76E003AT20 that message is received
+        _KACurtainSend("AT+SEND=ok");
```

Comment:
Just to avoid jumping between functions in loop. Plus, re-use Send function to compose message.